src/main/java/duke/TaskList.java

```diff
@@ -149,85 +147,83 @@ void addEvent(String[] tokens, Ui ui) throws DukeException {
         try {
             int listIndex = Integer.parseInt(tokens[1])-1;
             tasks.get(listIndex).setStatus("X");
-            ui.displayMessage("Nice! I've marked this task as done:\n" +
-                    tasks.get(listIndex).toString() + "\n");
+            return tasks.get(listIndex);
         } catch (NumberFormatException e) {
-            ui.displayMessage("Please specify a numerical task index to mark\n");
+            throw new DukeException("Please specify a numerical task index to mark\n");
         } catch (IndexOutOfBoundsException e) {
-            ui.displayMessage("Please specify a valid index to mark\n");
+            throw new DukeException("Please specify a valid index to mark\n");
         }
     }
 
     /**
-     * Method to unmark a list item and trigger corresponding <code>Ui</code> event.
+     * Unmarks specified <code>Task</code> in list.
      * @param tokens <code>String[]</code> of arguments from <code>Parser</code>, specifying the index to unmark.
      * @param ui Instance of <code>Ui</code> associated with the calling instance of <code>Duke</code>.
+     * @return The unmarked <code>Task</code>.
      */
-    void unmarkListItem(String[] tokens, Ui ui) {
+    Task unmarkListItem(String[] tokens, UiController ui) throws DukeException {
         try {
             int listIndex = Integer.parseInt(tokens[1]) - 1;
             tasks.get(listIndex).setStatus(" ");
-            ui.displayMessage("OK, I've marked this task as not done yet:\n" +
-                    tasks.get(listIndex).toString() + "\n");
+            return tasks.get(listIndex);
         } catch (NumberFormatException e) {
-            ui.displayMessage("Please specify a numerical task index to unmark\n");
+            throw new DukeException("Please specify a numerical task index to unmark\n");
         } catch (IndexOutOfBoundsException e) {
-            ui.displayMessage("Please specify a valid index to unmark\n");
+            throw new DukeException("Please specify a valid index to unmark\n");
         }
     }
 
     /**
-     * Method to delete an item from the list and trigger corresponding <code>Ui</code> event.
+     * Deletes specified <code>Task</code> from list.
      * @param tokens tokens <code>String[]</code> of arguments from <code>Parser</code>, specifying the index to delete.
      * @param ui Instance of <code>Ui</code> associated with the calling instance of <code>Duke</code>.
      * @throws DukeException In the event that the specified list index is out of bounds, or the argument corresponding
      * to the deletion index is not an integer.
+     * @return The deleted <code>Task</code>.
      */
-    void deleteItem(String[] tokens, Ui ui) throws DukeException {
+    Task deleteItem(String[] tokens, UiController ui) throws DukeException {
         if (tokens.length != 2) {
             throw new DukeException("please specify delete command as delete [list index]");
```

Comment:
Good to see that you've made many exceptions tailoring to different types of user inputs👍